File: CodeChef/FebruaryCodingChef/PrimeGame.cpp
// track all the prime numbers upto 1e6 + 5

#include<bits/stdc++.h>
#define int long long int
#define fastio ios_base::sync_with_stdio(false);cin.tie(0);
using namespace std;

const int N = 1e6 + 5;
int result[N];

void sieveOfEras() {
	vector<int> trackPrime;
	bool prime[N];
	memset(prime, true, sizeof(prime));

	for (int i=2; i*i < N; i++) {
		if (prime[i] == true) {
			for (int j=i*i; j<N; j+=i) {
				prime[j] = false;
			}
		}
	}
	for (int i=2; i<N; i++) {
		if (prime[i]) {
			trackPrime.push_back(i);
		}
	}
	auto first = trackPrime.begin(); // The auto keyword specifies that the type of the variable that is being declared will be automatically deducted from its initializer
	int count = 0;
	for (int i=0; i<N; i++) {
		if (*first <= i) {
			count ++;
			first ++;
		}
		result[i] = count;
	}
}

signed main() {
	fastio;
	sieveOfEras();
	int test;
	cin>>test;
	while (test --) {
		int x;
		int y;
		cin>>x>>y;
		if(result[x] <= y)
            cout<<"Chef"<<"\n";
        else
            cout<<"Divyam"<<"\n";
	}
	return 0;
}

File: CodeChef/FebruaryCodingChef/PrimeGame.java
import java.util.*;

public class PrimeGame {

	private static final int N = (int)1e6 + 5;
	private static int [] result = new int [N];
	public static String whoWins(int x, int y) {
		if (x == 1) {
			return "Chef";
		}
		List<Integer> list = countPrime();
		int count = 0;
		int first = list.get(0);
		for (int i=0; i<N; i++) {
			if (first <= i) {
				count ++;
				first ++;
			}
			result[i] = count;
		}
		return result[x] <= y ? "Chef" : "Divyam";

	}

	public static List<Integer> countPrime() {
		List<Integer> list = new ArrayList<>();
		int count = 0;
		boolean [] trackPrime = new boolean[N];
		for (int i=2; i*i<N; i++) {
			if (!trackPrime[i]) {
				for (int j=2; j*i<N; j++) {
					trackPrime[i * j] = true;
				}
			}
		}
		for (int i=2; i<N; i++) {
			if (!trackPrime[i])
				list.add(i);
		}
		return list;
	}

	

	public static void main (String [] args) {
		Scanner sc = new Scanner(System.in);
		int test = sc.nextInt();
		while (test-- > 0) {
			int x = sc.nextInt();
			int y = sc.nextInt();
			System.out.println(whoWins(x, y));
		}
	}
}